File: 백준/Silver/1992. 쿼드트리/쿼드트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

	static int N;
	static int[][] video;
	static StringBuilder sb = new StringBuilder();
	
	private static boolean check(int n, int x, int y) {
		for (int i = x; i < x + n; i++) {
			for (int j = y; j < y + n; j++) {
				if (video[x][y] != video[i][j]) return false;
			}
		}
		return true;
	}
	
	private static void solve(int n, int x, int y) {
		if (check(n, x, y)) {
			sb.append(video[x][y]);
			return;
		}
		
		sb.append("(");
		int mid = n / 2;
		for (int i = 0; i < 2; i++) {
			for (int j = 0; j < 2; j++) {
				solve(mid, x + i * mid, y + j * mid);
			}
		}
		sb.append(")");
	}
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		N = Integer.parseInt(br.readLine());
		
		video = new int[N][N];
		for (int i = 0; i < N; i++) {
			String temp = br.readLine();
			for (int j = 0; j < N; j++) {
				video[i][j] = temp.charAt(j) - '0';
			}
		}
		
		solve(N, 0, 0);
		System.out.println(sb);
	}

}

File: 백준/Silver/1992. 쿼드트리/쿼드트리.py
from sys import stdin
input = stdin.readline

def check(x, y, n):
    for i in range(x, x + n):
        for j in range(y, y + n):
            if video[x][y] != video[i][j]:
                return 0
    return 1

def solve(x, y, n):
    global answer
    if check(x, y, n):
        answer += video[x][y]
        return
    t = n // 2
    answer += '('
    for i in range(2):
        for j in range(2):
            solve(x + i * t, y + j * t, t)
    answer += ')'

N = int(input())
video = [list(input().strip()) for _ in range(N)]

answer = ''
solve(0, 0, N)
print(answer)